File: ringbuf/ringbuf.go
package ringbuf

import (
	"errors"
	"fmt"
	"io"
	"net"
	"os"
	"sync"
	"sync/atomic"
	"time"
	"unsafe"
)

const (
	StateEmpty   uint32 = 0
	StateWritten uint32 = 1
)

const (
	RoleHost   = "Host"
	RolePlugin = "Plugin"
)

type ringBuffer struct {
	data         []byte
	numSlots     int
	slotSize     int
	slotDataSize int
	writeIndex   int
	readIndex    int
	sysMapping   interface{}
}

// newRingBuffer creates or opens a file-backed mmap ring buffer and closes the file descriptor immediately.
func newRingBuffer(filePath string, numSlots int, slotDataSize int) (*ringBuffer, error) {
	if numSlots <= 0 || slotDataSize <= 0 {
		return nil, errors.New("invalid dimensions")
	}

	// Slot layout:
	// 0..3: flag (uint32)
	// 4..7: len (uint32)
	// 8..15: seq (uint64)
	// 16..19: writerWaiting (uint32)
	// 20..23: readerWaiting (uint32)
	// 24..: payload (slotDataSize)
	slotSize := 24 + slotDataSize
	totalSize := numSlots * slotSize

	data, sysMapping, err := mmapFile(filePath, totalSize)
	if err != nil {
		return nil, err
	}

	return &ringBuffer{
		data:         data,
		numSlots:     numSlots,
		slotSize:     slotSize,
		slotDataSize: slotDataSize,
		writeIndex:   0,
		readIndex:    0,
		sysMapping:   sysMapping,
	}, nil
}

// Close unmaps the memory mapped region.
func (rb *ringBuffer) Close() error {
	if rb.data != nil {
		if err := unmapFile(rb.data, rb.sysMapping); err != nil {
			return fmt.Errorf("failed to unmap: %w", err)
		}
		rb.data = nil
	}
	return nil
}

// Clear resets the header flags of all slots to StateEmpty.
func (rb *ringBuffer) Clear() {
	for i := 0; i < rb.numSlots; i++ {
		atomic.StoreUint32(rb.getFlagPtr(i), StateEmpty)
		atomic.StoreUint32(rb.getLenPtr(i), 0)
		atomic.StoreUint64(rb.getSeqPtr(i), 0)
		atomic.StoreUint32(rb.getWriterWaitingPtr(i), 0)
		atomic.StoreUint32(rb.getReaderWaitingPtr(i), 0)
	}
	rb.writeIndex = 0
	rb.readIndex = 0
}

// Write writes payload and sequence to the current slot if empty.
// If not empty, it blocks on UDS connection read waiting for an empty slot signal.
func (rb *ringBuffer) Write(payload []byte, seq uint64, timeout time.Duration, sigConn net.Conn) error {
	if len(payload) > rb.slotDataSize {
		return fmt.Errorf("payload size %d exceeds slot capacity %d", len(payload), rb.slotDataSize)
	}

	flagPtr := rb.getFlagPtr(rb.writeIndex)
	writerWaitingPtr := rb.getWriterWaitingPtr(rb.writeIndex)
	start := time.Now()

	for atomic.LoadUint32(flagPtr) != StateEmpty {
		// Mark that this writer is waiting on this slot
		atomic.StoreUint32(writerWaitingPtr, 1)

		// Double-check after setting waiting flag to prevent race condition (lost wakeup)
		if atomic.LoadUint32(flagPtr) == StateEmpty {
			atomic.StoreUint32(writerWaitingPtr, 0)
			break
		}

		if timeout > 0 {
			sigConn.SetReadDeadline(time.Now().Add(timeout - time.Since(start)))
		} else {
			sigConn.SetReadDeadline(time.Time{})
		}
		var signalBuf [1]byte
		_, err := sigConn.Read(signalBuf[:])
		
		atomic.StoreUint32(writerWaitingPtr, 0)

		if err != nil {
			idx := rb.writeIndex
			rb.writeIndex = (rb.writeIndex + 1) % rb.numSlots
			return fmt.Errorf("write timeout on slot %d waiting for empty signal: %w", idx, err)
		}
	}

	// Copy payload
	dest := rb.getPayloadSlice(rb.writeIndex)
	copy(dest, payload)

	// Store length and sequence
	atomic.StoreUint32(rb.getLenPtr(rb.writeIndex), uint32(len(payload)))
	atomic.StoreUint64(rb.getSeqPtr(rb.writeIndex), seq)

	// Atomically mark as written
	atomic.StoreUint32(flagPtr, StateWritten)

	// Signal the reader if they are waiting
	readerWaitingPtr := rb.getReaderWaitingPtr(rb.writeIndex)
	if atomic.CompareAndSwapUint32(readerWaitingPtr, 1, 0) {
		var token = [1]byte{0x01}
		sigConn.SetWriteDeadline(time.Now().Add(1 * time.Second))
		sigConn.Write(token[:])
	}

	rb.writeIndex = (rb.writeIndex + 1) % rb.numSlots
	return nil
}

// Read waits for the current slot to be written.
// If empty, it blocks on UDS connection read waiting for a written slot signal.
func (rb *ringBuffer) Read(timeout time.Duration, sigConn net.Conn) ([]byte, uint64, error) {
	flagPtr := rb.getFlagPtr(rb.readIndex)
	readerWaitingPtr := rb.getReaderWaitingPtr(rb.readIndex)
	start := time.Now()

	for atomic.LoadUint32(flagPtr) != StateWritten {
		// Mark that this reader is waiting on this slot
		atomic.StoreUint32(readerWaitingPtr, 1)

		// Double-check after setting waiting flag to prevent race condition (lost wakeup)
		if atomic.LoadUint32(flagPtr) == StateWritten {
			atomic.StoreUint32(readerWaitingPtr, 0)
			break
		}

		if timeout > 0 {
			sigConn.SetReadDeadline(time.Now().Add(timeout - time.Since(start)))
		} else {
			sigConn.SetReadDeadline(time.Time{})
		}
		var signalBuf [1]byte
		_, err := sigConn.Read(signalBuf[:])

		atomic.StoreUint32(readerWaitingPtr, 0)

		if err != nil {
			return nil, 0, fmt.Errorf("read timeout on slot %d waiting for write signal: %w", rb.readIndex, err)
		}
	}

	length := atomic.LoadUint32(rb.getLenPtr(rb.readIndex))
	seq := atomic.LoadUint64(rb.getSeqPtr(rb.readIndex))

	if int(length) > rb.slotDataSize {
		return nil, 0, fmt.Errorf("invalid payload length %d on slot %d", length, rb.readIndex)
	}

	payload := make([]byte, length)
	src := rb.getPayloadSlice(rb.readIndex)
	copy(payload, src[:length])

	// Mark as read/empty
	atomic.StoreUint32(flagPtr, StateEmpty)

	// Signal the writer if they are waiting
	writerWaitingPtr := rb.getWriterWaitingPtr(rb.readIndex)
	if atomic.CompareAndSwapUint32(writerWaitingPtr, 1, 0) {
		var token = [1]byte{0x02}
		sigConn.SetWriteDeadline(time.Now().Add(1 * time.Second))
		sigConn.Write(token[:])
	}

	rb.readIndex = (rb.readIndex + 1) % rb.numSlots
	return payload, seq, nil
}

// GetWriteIndex returns the current write index of the buffer slot being accessed.
func (rb *ringBuffer) GetWriteIndex() int {
	return rb.writeIndex
}

// GetReadIndex returns the current read index of the buffer slot being accessed.
func (rb *ringBuffer) GetReadIndex() int {
	return rb.readIndex
}

func (rb *ringBuffer) getFlagPtr(slot int) *uint32 {
	offset := slot * rb.slotSize
	return (*uint32)(unsafe.Pointer(&rb.data[offset]))
}

func (rb *ringBuffer) getLenPtr(slot int) *uint32 {
	offset := slot * rb.slotSize
	return (*uint32)(unsafe.Pointer(&rb.data[offset+4]))
}

func (rb *ringBuffer) getSeqPtr(slot int) *uint64 {
	offset := slot * rb.slotSize
	return (*uint64)(unsafe.Pointer(&rb.data[offset+8]))
}

func (rb *ringBuffer) getWriterWaitingPtr(slot int) *uint32 {
	offset := slot * rb.slotSize
	return (*uint32)(unsafe.Pointer(&rb.data[offset+16]))
}

func (rb *ringBuffer) getReaderWaitingPtr(slot int) *uint32 {
	offset := slot * rb.slotSize
	return (*uint32)(unsafe.Pointer(&rb.data[offset+20]))
}

func (rb *ringBuffer) getPayloadSlice(slot int) []byte {
	offset := slot * rb.slotSize
	return rb.data[offset+24 : offset+24+rb.slotDataSize]
}

type Writer struct {
	rb           *ringBuffer
	sigConn      net.Conn
	writeTimeout time.Duration
	nextWriteSeq uint64
	writeMu      sync.Mutex
}

// Write implements io.Writer. It partitions p into slots and writes them.
func (w *Writer) Write(p []byte) (n int, err error) {
	w.writeMu.Lock()
	defer w.writeMu.Unlock()

	total := len(p)
	written := 0

	for len(p) > 0 {
		chunkSize := len(p)
		if chunkSize > w.rb.slotDataSize {
			chunkSize = w.rb.slotDataSize
		}

		err = w.rb.Write(p[:chunkSize], w.nextWriteSeq, w.writeTimeout, w.sigConn)
		if err != nil {
			return written, err
		}

		w.nextWriteSeq++
		written += chunkSize
		p = p[chunkSize:]
	}

	return total, nil
}

// Close closes the signaling connection and unmaps the writer memory.
func (w *Writer) Close() error {
	var errs []error
	if w.sigConn != nil {
		if err := w.sigConn.Close(); err != nil {
			errs = append(errs, err)
		}
		w.sigConn = nil
	}
	if w.rb != nil {
		if err := w.rb.Close(); err != nil {
			errs = append(errs, err)
		}
		w.rb = nil
	}
	if len(errs) > 0 {
		return errs[0]
	}
	return nil
}

type Reader struct {
	rb          *ringBuffer
	sigConn     net.Conn
	readTimeout time.Duration
	readBuf     []byte
	readMu      sync.Mutex
}

// Read implements io.Reader. It reads next slot when internal buffer is empty.
func (r *Reader) Read(p []byte) (n int, err error) {
	r.readMu.Lock()
	defer r.readMu.Unlock()

	if len(p) == 0 {
		return 0, nil
	}

	if len(r.readBuf) == 0 {
		payload, _, err := r.rb.Read(r.readTimeout, r.sigConn)
		if err != nil {
			return 0, err
		}
		r.readBuf = payload
	}

	n = copy(p, r.readBuf)
	r.readBuf = r.readBuf[n:]
	return n, nil
}

// ReadFull reads exactly len(buf) bytes from the Reader into buf.
// It blocks until all bytes are read, or an error/timeout occurs.
func (r *Reader) ReadFull(buf []byte) (n int, err error) {
	return io.ReadFull(r, buf)
}

// Close closes the signaling connection and unmaps the reader memory.
func (r *Reader) Close() error {
	var errs []error
	if r.sigConn != nil {
		if err := r.sigConn.Close(); err != nil {
			errs = append(errs, err)
		}
		r.sigConn = nil
	}
	if r.rb != nil {
		if err := r.rb.Close(); err != nil {
			errs = append(errs, err)
		}
		r.rb = nil
	}
	if len(errs) > 0 {
		return errs[0]
	}
	return nil
}

// NewConnection creates the Writer and Reader pair using the base filepath and node role.
// It also establishes UDS signaling connections over `{basePath}_sig.sock`.
func NewConnection(basePath string, role string, numSlots, slotDataSize int, writeTimeout, readTimeout time.Duration) (*Writer, *Reader, error) {
	var writePath, readPath string
	sigSockPath := basePath + "_sig.sock"

	var conn1, conn2 net.Conn

	if role == RoleHost {
		writePath = basePath + "_writer"
		readPath = basePath + "_reader"

		_ = os.Remove(sigSockPath)
		ln, err := net.Listen("unix", sigSockPath)
		if err != nil {
			return nil, nil, fmt.Errorf("host failed to listen on signaling socket: %w", err)
		}
		defer ln.Close()

		// Accept connection 1: Host Writer <-> Plugin Reader
		conn1, err = ln.Accept()
		if err != nil {
			return nil, nil, fmt.Errorf("host failed to accept first sig connection: %w", err)
		}

		// Accept connection 2: Plugin Writer <-> Host Reader
		conn2, err = ln.Accept()
		if err != nil {
			conn1.Close()
			return nil, nil, fmt.Errorf("host failed to accept second sig connection: %w", err)
		}

	} else if role == RolePlugin {
		writePath = basePath + "_reader"
		readPath = basePath + "_writer"

		var err error
		// Dial connection 1: Host Writer <-> Plugin Reader
		for attempt := 0; attempt < 100; attempt++ {
			conn1, err = net.Dial("unix", sigSockPath)
			if err == nil {
				break
			}
			time.Sleep(10 * time.Millisecond)
		}
		if err != nil {
			return nil, nil, fmt.Errorf("plugin failed to dial first sig connection: %w", err)
		}

		// Dial connection 2: Plugin Writer <-> Host Reader
		for attempt := 0; attempt < 100; attempt++ {
			conn2, err = net.Dial("unix", sigSockPath)
			if err == nil {
				break
			}
			time.Sleep(10 * time.Millisecond)
		}
		if err != nil {
			conn1.Close()
			return nil, nil, fmt.Errorf("plugin failed to dial second sig connection: %w", err)
		}

	} else {
		return nil, nil, fmt.Errorf("invalid role: %s, must be 'Host' or 'Plugin'", role)
	}

	writeRB, err := newRingBuffer(writePath, numSlots, slotDataSize)
	if err != nil {
		conn1.Close()
		conn2.Close()
		return nil, nil, err
	}
	writeRB.Clear() // Clear write flags for a fresh start

	readRB, err := newRingBuffer(readPath, numSlots, slotDataSize)
	if err != nil {
		writeRB.Close()
		conn1.Close()
		conn2.Close()
		return nil, nil, err
	}

	var writer *Writer
	var reader *Reader

	if role == RoleHost {
		writer = &Writer{
			rb:           writeRB,
			sigConn:      conn1, // Host Writer uses conn1 to signal Plugin Reader
			writeTimeout: writeTimeout,
		}
		reader = &Reader{
			rb:          readRB,
			sigConn:     conn2, // Host Reader uses conn2 to read Plugin Writer signals
			readTimeout: readTimeout,
		}
	} else {
		// role == RolePlugin
		writer = &Writer{
			rb:           writeRB,
			sigConn:      conn2, // Plugin Writer uses conn2 to signal Host Reader
			writeTimeout: writeTimeout,
		}
		reader = &Reader{
			rb:          readRB,
			sigConn:     conn1, // Plugin Reader uses conn1 to read Host Writer signals
			readTimeout: readTimeout,
		}
	}

	return writer, reader, nil
}

File: typescript/ringbuf-ts/src/ringbuf.ts
import * as fs from 'fs';
import * as net from 'net';

export const STATE_EMPTY = 0;
export const STATE_WRITTEN = 1;

export const ROLE_HOST = "Host";
export const ROLE_PLUGIN = "Plugin";

export class TimeoutError extends Error {
    constructor(message: string) {
        super(message);
        this.name = "TimeoutError";
    }
}

export class RingBuffer {
    public fd: number;
    public numSlots: number;
    public slotDataSize: number;
    public slotSize: number;
    public totalSize: number;
    public writeIndex = 0;
    public readIndex = 0;

    constructor(filePath: string, numSlots: number, slotDataSize: number) {
        if (numSlots <= 0 || slotDataSize <= 0) {
            throw new Error("invalid dimensions");
        }
        this.numSlots = numSlots;
        this.slotDataSize = slotDataSize;
        this.slotSize = 24 + slotDataSize;
        this.totalSize = numSlots * this.slotSize;

        try {
            this.fd = fs.openSync(filePath, 'r+');
        } catch (e) {
            this.fd = fs.openSync(filePath, 'w+');
        }
        fs.ftruncateSync(this.fd, this.totalSize);
    }

    close(): void {
        if (this.fd !== -1) {
            try {
                fs.closeSync(this.fd);
            } catch (e) {}
            this.fd = -1;
        }
    }

    clear(): void {
        for (let i = 0; i < this.numSlots; i++) {
            this.setFlag(i, STATE_EMPTY);
            this.setLen(i, 0);
            this.setSeq(i, 0n);
            this.setWriterWaiting(i, 0);
            this.setReaderWaiting(i, 0);
        }
        this.writeIndex = 0;
        this.readIndex = 0;
    }

    private _offset(slot: number): number {
        return slot * this.slotSize;
    }

    getFlag(slot: number): number {
        const buf = Buffer.alloc(4);
        fs.readSync(this.fd, buf, 0, 4, this._offset(slot));
        return buf.readUInt32LE(0);
    }

    setFlag(slot: number, val: number): void {
        const buf = Buffer.alloc(4);
        buf.writeUInt32LE(val, 0);
        fs.writeSync(this.fd, buf, 0, 4, this._offset(slot));
    }

    getLen(slot: number): number {
        const buf = Buffer.alloc(4);
        fs.readSync(this.fd, buf, 0, 4, this._offset(slot) + 4);
        return buf.readUInt32LE(0);
    }

    setLen(slot: number, val: number): void {
        const buf = Buffer.alloc(4);
        buf.writeUInt32LE(val, 0);
        fs.writeSync(this.fd, buf, 0, 4, this._offset(slot) + 4);
    }

    getSeq(slot: number): bigint {
        const buf = Buffer.alloc(8);
        fs.readSync(this.fd, buf, 0, 8, this._offset(slot) + 8);
        return buf.readBigUInt64LE(0);
    }

    setSeq(slot: number, val: bigint | number): void {
        const buf = Buffer.alloc(8);
        buf.writeBigUInt64LE(BigInt(val), 0);
        fs.writeSync(this.fd, buf, 0, 8, this._offset(slot) + 8);
    }

    getWriterWaiting(slot: number): number {
        const buf = Buffer.alloc(4);
        fs.readSync(this.fd, buf, 0, 4, this._offset(slot) + 16);
        return buf.readUInt32LE(0);
    }

    setWriterWaiting(slot: number, val: number): void {
        const buf = Buffer.alloc(4);
        buf.writeUInt32LE(val, 0);
        fs.writeSync(this.fd, buf, 0, 4, this._offset(slot) + 16);
    }

    getReaderWaiting(slot: number): number {
        const buf = Buffer.alloc(4);
        fs.readSync(this.fd, buf, 0, 4, this._offset(slot) + 20);
        return buf.readUInt32LE(0);
    }

    setReaderWaiting(slot: number, val: number): void {
        const buf = Buffer.alloc(4);
        buf.writeUInt32LE(val, 0);
        fs.writeSync(this.fd, buf, 0, 4, this._offset(slot) + 20);
    }

    getPayload(slot: number, len: number): Buffer {
        const buf = Buffer.alloc(len);
        fs.readSync(this.fd, buf, 0, len, this._offset(slot) + 24);
        return buf;
    }

    setPayload(slot: number, payload: Buffer): void {
        fs.writeSync(this.fd, payload, 0, payload.length, this._offset(slot) + 24);
    }

    async write(payload: Buffer, seq: bigint | number, timeoutMs: number, sigMgr: SocketSignalManager): Promise<void> {
        if (payload.length > this.slotDataSize) {
            throw new Error(`payload size ${payload.length} exceeds slot capacity ${this.slotDataSize}`);
        }

        const start = Date.now();
        while (this.getFlag(this.writeIndex) !== STATE_EMPTY) {
            this.setWriterWaiting(this.writeIndex, 1);

            // Yield to execute a memory barrier
            await new Promise(r => setTimeout(r, 0));

            if (this.getFlag(this.writeIndex) === STATE_EMPTY) {
                this.setWriterWaiting(this.writeIndex, 0);
                break;
            }

            const remaining = timeoutMs > 0 ? timeoutMs - (Date.now() - start) : 0;
            if (timeoutMs > 0 && remaining <= 0) {
                this.setWriterWaiting(this.writeIndex, 0);
                const idx = this.writeIndex;
                this.writeIndex = (this.writeIndex + 1) % this.numSlots;
                throw new TimeoutError(`write timeout on slot ${idx} waiting for empty signal`);
            }

            try {
                await sigMgr.waitWrite(remaining);
            } catch (err) {
                this.setWriterWaiting(this.writeIndex, 0);
                const idx = this.writeIndex;
                this.writeIndex = (this.writeIndex + 1) % this.numSlots;
                throw err;
            }

            this.setWriterWaiting(this.writeIndex, 0);
        }

        const slotIdx = this.writeIndex;
        this.setPayload(slotIdx, payload);
        this.setLen(slotIdx, payload.length);
        this.setSeq(slotIdx, seq);

        this.setFlag(slotIdx, STATE_WRITTEN);

        if (this.getReaderWaiting(slotIdx) === 1) {
            this.setReaderWaiting(slotIdx, 0);
            try {
                sigMgr.send(0x01);
            } catch (e) {}
        }

        this.writeIndex = (this.writeIndex + 1) % this.numSlots;
    }

    async read(timeoutMs: number, sigMgr: SocketSignalManager): Promise<[Buffer, bigint]> {
        const start = Date.now();
        while (this.getFlag(this.readIndex) !== STATE_WRITTEN) {
            this.setReaderWaiting(this.readIndex, 1);

            // Yield to execute a memory barrier
            await new Promise(r => setTimeout(r, 0));

            if (this.getFlag(this.readIndex) === STATE_WRITTEN) {
                this.setReaderWaiting(this.readIndex, 0);
                break;
            }

            const remaining = timeoutMs > 0 ? timeoutMs - (Date.now() - start) : 0;
            if (timeoutMs > 0 && remaining <= 0) {
                this.setReaderWaiting(this.readIndex, 0);
                throw new TimeoutError(`read timeout on slot ${this.readIndex} waiting for write signal`);
            }

            try {
                await sigMgr.waitRead(remaining);
            } catch (err) {
                this.setReaderWaiting(this.readIndex, 0);
                throw err;
            }

            this.setReaderWaiting(this.readIndex, 0);
        }

        const slotIdx = this.readIndex;
        const len = this.getLen(slotIdx);
        const seq = this.getSeq(slotIdx);

        if (len > this.slotDataSize) {
            throw new Error(`invalid payload length ${len} on slot ${slotIdx}`);
        }

        const payload = this.getPayload(slotIdx, len);

        this.setFlag(slotIdx, STATE_EMPTY);

        if (this.getWriterWaiting(slotIdx) === 1) {
            this.setWriterWaiting(slotIdx, 0);
            try {
                sigMgr.send(0x02);
            } catch (e) {}
        }

        this.readIndex = (this.readIndex + 1) % this.numSlots;
        return [payload, seq];
    }
}

export class SocketSignalManager {
    private socket: net.Socket;
    private writeWaiters: (() => void)[] = [];
    private readWaiters: (() => void)[] = [];

    constructor(socket: net.Socket) {
        this.socket = socket;
        this.socket.on('error', () => {}); // Prevent unhandled 'error' crash on ECONNRESET
        this.socket.on('data', (chunk) => {
            for (const byte of chunk) {
                if (byte === 0x01) {
                    const cb = this.readWaiters.shift();
                    if (cb) cb();
                } else if (byte === 0x02) {
                    const cb = this.writeWaiters.shift();
                    if (cb) cb();
                }
            }
        });
    }

    waitWrite(timeoutMs: number): Promise<void> {
        return new Promise<void>((resolve, reject) => {
            let timer: NodeJS.Timeout | null = null;
            const cb = () => {
                if (timer) clearTimeout(timer);
                resolve();
            };
            this.writeWaiters.push(cb);

            if (timeoutMs > 0) {
                timer = setTimeout(() => {
                    const idx = this.writeWaiters.indexOf(cb);
                    if (idx !== -1) {
                        this.writeWaiters.splice(idx, 1);
                    }
                    reject(new TimeoutError("write timeout"));
                }, timeoutMs);
            }
        });
    }

    waitRead(timeoutMs: number): Promise<void> {
        return new Promise<void>((resolve, reject) => {
            let timer: NodeJS.Timeout | null = null;
            const cb = () => {
                if (timer) clearTimeout(timer);
                resolve();
            };
            this.readWaiters.push(cb);

            if (timeoutMs > 0) {
                timer = setTimeout(() => {
                    const idx = this.readWaiters.indexOf(cb);
                    if (idx !== -1) {
                        this.readWaiters.splice(idx, 1);
                    }
                    reject(new TimeoutError("read timeout"));
                }, timeoutMs);
            }
        });
    }

    send(token: number): void {
        this.socket.write(Buffer.from([token]));
    }

    close(): void {
        this.socket.end();
    }
}

class Mutex {
    private queue: (() => void)[] = [];
    private locked = false;

    acquire(): Promise<() => void> {
        return new Promise((resolve) => {
            const release = () => {
                if (this.queue.length > 0) {
                    const next = this.queue.shift();
                    if (next) next();
                } else {
                    this.locked = false;
                }
            };

            if (this.locked) {
                this.queue.push(() => resolve(release));
            } else {
                this.locked = true;
                resolve(release);
            }
        });
    }
}

export class Writer {
    private rb: RingBuffer;
    private sigMgr: SocketSignalManager;
    private writeTimeoutMs: number;
    private nextWriteSeq = 0n;
    private writeMutex = new Mutex();

    constructor(rb: RingBuffer, sigMgr: SocketSignalManager, writeTimeoutMs: number) {
        this.rb = rb;
        this.sigMgr = sigMgr;
        this.writeTimeoutMs = writeTimeoutMs;
    }

    async write(p: Buffer): Promise<number> {
        const release = await this.writeMutex.acquire();
        try {
            const total = p.length;
            let offset = 0;
            while (offset < total) {
                const chunkSize = Math.min(total - offset, this.rb.slotDataSize);
                const chunk = p.subarray(offset, offset + chunkSize);
                await this.rb.write(chunk, this.nextWriteSeq, this.writeTimeoutMs, this.sigMgr);
                this.nextWriteSeq++;
                offset += chunkSize;
            }
            return total;
        } finally {
            release();
        }
    }

    close(): void {
        this.sigMgr.close();
        this.rb.close();
    }
}

export class Reader {
    private rb: RingBuffer;
    private sigMgr: SocketSignalManager;
    private readTimeoutMs: number;
    private readBuf: Buffer = Buffer.alloc(0);
    private readMutex = new Mutex();

    constructor(rb: RingBuffer, sigMgr: SocketSignalManager, readTimeoutMs: number) {
        this.rb = rb;
        this.sigMgr = sigMgr;
        this.readTimeoutMs = readTimeoutMs;
    }

    async read(n: number): Promise<Buffer> {
        const release = await this.readMutex.acquire();
        try {
            if (n <= 0) return Buffer.alloc(0);
            if (this.readBuf.length === 0) {
                const [payload] = await this.rb.read(this.readTimeoutMs, this.sigMgr);
                this.readBuf = payload;
            }

            const chunkSize = Math.min(n, this.readBuf.length);
            const res = this.readBuf.subarray(0, chunkSize);
            this.readBuf = this.readBuf.subarray(chunkSize);
            return Buffer.from(res);
        } finally {
            release();
        }
    }

    async readFull(n: number): Promise<Buffer> {
        const chunks: Buffer[] = [];
        let bytesRead = 0;
        while (bytesRead < n) {
            const chunk = await this.read(n - bytesRead);
            if (chunk.length === 0) {
                throw new Error("connection closed during readFull");
            }
            chunks.push(chunk);
            bytesRead += chunk.length;
        }
        return Buffer.concat(chunks);
    }

    close(): void {
        this.sigMgr.close();
        this.rb.close();
    }
}

export async function newConnection(
    basePath: string,
    role: string,
    numSlots: number,
    slotDataSize: number,
    writeTimeoutMs: number,
    readTimeoutMs: number
): Promise<[Writer, Reader]> {
    const sigSockPath = basePath + "_sig.sock";

    const writePath = role === ROLE_HOST ? basePath + "_writer" : basePath + "_reader";
    const readPath = role === ROLE_HOST ? basePath + "_reader" : basePath + "_writer";

    // Instantiate and clear shared memory ring buffers before connection accepted to prevent race conditions
    const writeRb = new RingBuffer(writePath, numSlots, slotDataSize);
    writeRb.clear();
    const readRb = new RingBuffer(readPath, numSlots, slotDataSize);

    let socket1: net.Socket;
    let socket2: net.Socket;

    if (role === ROLE_HOST) {
        if (fs.existsSync(sigSockPath)) {
            try {
                fs.unlinkSync(sigSockPath);
            } catch (e) {}
        }

        const server = net.createServer();
        server.listen(sigSockPath);

        const socketsPromise = new Promise<[net.Socket, net.Socket]>((resolve, reject) => {
            const accepted: net.Socket[] = [];
            server.on('connection', (socket) => {
                accepted.push(socket);
                if (accepted.length === 2) {
                    server.close();
                    resolve([accepted[0], accepted[1]]);
                }
            });
            server.on('error', (err) => reject(err));
        });

        [socket1, socket2] = await socketsPromise;
    } else if (role === ROLE_PLUGIN) {
        const connect = async () => {
            for (let i = 0; i < 100; i++) {
                try {
                    const socket = net.createConnection(sigSockPath);
                    await new Promise<void>((resolve, reject) => {
                        socket.on('connect', resolve);
                        socket.on('error', reject);
                    });
                    return socket;
                } catch (e) {
                    await new Promise(r => setTimeout(r, 10));
                }
            }
            throw new Error("plugin failed to dial sig connection");
        };

        socket1 = await connect();
        socket2 = await connect();
    } else {
        throw new Error(`invalid role: ${role}`);
    }

    const sigMgr1 = new SocketSignalManager(socket1);
    const sigMgr2 = new SocketSignalManager(socket2);

    let writer: Writer;
    let reader: Reader;

    if (role === ROLE_HOST) {
        writer = new Writer(writeRb, sigMgr1, writeTimeoutMs);
        reader = new Reader(readRb, sigMgr2, readTimeoutMs);
    } else {
        writer = new Writer(writeRb, sigMgr2, writeTimeoutMs);
        reader = new Reader(readRb, sigMgr1, readTimeoutMs);
    }

    return [writer, reader];
}
